### PR TITLE
Use quantized nvfp4 in scaled mma op

### PR DESCRIPTION
**Pull request by protonu**
This adds a Python test to used quantized nvfp4 in the scaled mma op. Please note the relaxed accuracy checks in the test. For inferring allocation sizes, we skip the inputs to scaled_mm and output of block quantization.

### CODE REVIEW

**Comment by protonu** [comment]
!test

**Comment by jjsjann123** [comment]
@protonu Would you mind resolving the conflicts? I think that's coming from the other merged PR.

**Comment by protonu** [comment]
!test

File: csrc/tensor_metadata.cpp [hunk 1]
```diff
@@ -342,8 +342,25 @@ inferAndValidateAllocationSizesAndStrides(
     ExpressionEvaluator ee) {
   auto [allocation_sizes, allocation_strides] =
       inferAllocationSizesAndStrides(tensor, tv, ee);
-  // Only validate final sizes and strides when we have a non-empty tensor.
-  if (tensor.numel() != 0) {
+
+  // Skip validation for swizzled scales in ScaledMmaOp.
+  // We only want to skip the validation for the scales
+  // produced by the Block Quantization Op, but since we can't
```
Comment (jjsjann123) [inline csrc/tensor_metadata.cpp]: I *think* this check runs on segmented fusion as well. So if the segmentation happened right at the boundary of quantization, the output from block quantization op wouldn't skip the check, since we don't see the consumer in the fusion.

One more question regarding `We only want to skip the validation for the scales
Comment view  produced by the Block Quantization Op, but since we can't do that here`, why can't we do that here?

Comment (protonu) [reply csrc/tensor_metadata.cpp]: Thanks - updated the PR based on our offline conversations.

**Comment by protonu** [comment]
!test

File: csrc/tensor_metadata.cpp [hunk 2]
```diff
@@ -342,8 +342,27 @@ inferAndValidateAllocationSizesAndStrides(
...
+    auto bqop = tv->definition()->as<BlockQuantizationOp>();
+    if (bqop->isSwizzledScales() && tv == bqop->blockScales()) {
+      skip_validation = true;
+    }
+  }
+
+  // Skip validation for scale input to ScaledMmaOp as it will be swizzled.
+  // This check is a bit of an overkill as it will skip all inputs to the op.
+  if (tv->uses().size() == 1 && tv->uses().at(0)->isA<ScaledMmaOp>()) {
+    skip_validation = true;
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/tensor_metadata.cpp]: I agree with greptile review here.

[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/tensor_metadata.cpp]: Our code comment mentioned skip only the scale input, but the check didn't validate that.

File: tests/python/direct/test_narrow_precision.py [hunk 3]
```diff
@@ -67,6 +67,39 @@ def nvfp4_quantize_with_te(input_tensor):
...
+        tensor: Input tensor to compute scale for
+        device: Device to create constant tensors on
+
+    Returns:
+        Global scale as (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / max(abs(tensor))
+    """
+    amax = torch.max(torch.abs(tensor)).to(torch.float32)
+    float4_max = torch.tensor(6.0, device=device, dtype=torch.float32)
+    float8_max = torch.tensor(448.0, device=device, dtype=torch.float32)
+    return torch.div(float8_max * float4_max, amax)
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tests/python/direct/test_narrow_precision.py]: valid point to avoid div by zero.

File: tests/python/direct/test_narrow_precision.py [hunk 4]
```diff
@@ -67,6 +67,39 @@ def nvfp4_quantize_with_te(input_tensor):
...
+    Args:
+        tensor: Input tensor to compute scale for
+        device: Device to create constant tensors on
+
+    Returns:
+        Global scale as (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / max(abs(tensor))
+    """
+    amax = torch.max(torch.abs(tensor)).to(torch.float32)
+    float4_max = torch.tensor(6.0, device=device, dtype=torch.float32)
+    float8_max = torch.tensor(448.0, device=device, dtype=torch.float32)
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py]: Why do we want to take another `device` arg here? If `device != tensor.device`, the div later would fail?

File: tests/python/direct/test_narrow_precision.py [hunk 5]
```diff
@@ -211,6 +244,148 @@ def nvfuser_fusion_id0(fd: FusionDefinition) -> None:
...
+    mat1_ref = torch.randn((m, k), dtype=torch.bfloat16, device="cuda")
+    mat2_ref = torch.randn((n, k), dtype=torch.bfloat16, device="cuda")
+
+    # Quantize both matrices using Transformer Engine
+    mat1_quantized, mat1_scale_inv, global_sf1 = extract_te_nvfp4_metadata(mat1_ref)
+    mat2_quantized, mat2_scale_inv, global_sf2 = extract_te_nvfp4_metadata(mat2_ref)
+
+    # Convert to appropriate dtypes for scaled_mm
+    mat2_quantized = mat2_quantized.view(torch.float4_e2m1fn_x2)
+    mat2_scale_inv = mat2_scale_inv.view(torch.float8_e4m3fn)
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py]: nitpick: looks like we are only using f4x2 and f8 dtype from `extract_te_nvfp4_metadata`. Might as well put the cast inside that function.

File: tests/python/direct/test_narrow_precision.py [hunk 6]
```diff
@@ -211,6 +244,148 @@ def nvfuser_fusion_id0(fd: FusionDefinition) -> None:
...
+    abs_diff = torch.abs(outputs[0] - outputs_baseline[0])
+    max_diff = torch.max(abs_diff)
+    assert max_diff <= 10.0, f"Max difference {max_diff:.4f} exceeds threshold of 10.0"
+
+    # Check that large differences (> 5.0) are rare (< 10% of elements)
+    large_diff_count = torch.count_nonzero(torch.gt(abs_diff, 5.0))
+    large_diff_ratio = large_diff_count / abs_diff.numel()
+    assert (
+        large_diff_ratio < 0.1
+    ), f"Large diff ratio {large_diff_ratio:.2%} exceeds 10% threshold"
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py:385]: I don't know how to interpret the differences. Have we checked with TE folks about their validation for nvfp4 yet?

**Comment by protonu** [comment]
!test

**Comment by protonu** [comment]
!test

File: tests/python/direct/test_narrow_precision.py (hunk 6 above)
Comment (protonu) [reply tests/python/direct/test_narrow_precision.py:385]: I'll start a thread with them. Since the difference comes from the quantization and not the linear operation, I thought we could land this in parallel.

File: tests/python/direct/test_narrow_precision.py [hunk 7]
```diff
@@ -67,6 +67,47 @@ def nvfp4_quantize_with_te(input_tensor):
...
+        Global scale as (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / max(abs(tensor))
+    """
+    amax = torch.max(torch.abs(tensor)).to(torch.float32)
+    x_global_scale = (FLOAT8_E4M3_MAX * FLOAT4_E2M1_MAX) / amax
+    x_global_scale = torch.clamp(x_global_scale, max=torch.finfo(torch.float32).max)
+
+    # Handle edge cases: zero input or invalid scale
+    invalid_mask = (amax == 0.0) | (x_global_scale == 0.0)
+    x_global_scale = torch.where(
+        invalid_mask, torch.tensor(1.0, device="cuda"), x_global_scale
```
Comment (jjsjann123) [inline tests/python/direct/test_narrow_precision.py]: super nitpick: we can just use `tensor.device` to avoid device mismatch, which isn't a problem for scalar tensor... so no big deal.

Also, `torch.where(invalid_mask, 1.0, x_global_scale)` should work?!

**Comment by protonu** [comment]
!test
